src/ripple/app/tx/impl/Change.cpp

```diff
@@ -124,14 +125,66 @@ Change::applyAmendment()
             amendment) != amendments.end ())
         return tefALREADY;
 
-    amendments.push_back (amendment);
-    amendmentObject->setFieldV256 (sfAmendments, amendments);
-    view().update (amendmentObject);
+    auto flags = mTxn.getFlags ();
+
+    const bool gotMajority = (flags & tfGotMajority) != 0;
+    const bool lostMajority = (flags & tfLostMajority) != 0;
+
+    if (gotMajority && lostMajority)
+        return temINVALID_FLAG;
 
-    getApp().getAmendmentTable ().enable (amendment);
+    STArray newMajorities (sfMajorities);
 
-    if (!getApp().getAmendmentTable ().isSupported (amendment))
-        getApp().getOPs ().setAmendmentBlocked ();
+    bool found = false;
+    if (amendmentObject->isFieldPresent (sfMajorities))
+    {
+        const STArray &oldMajorities = amendmentObject->getFieldArray (sfMajorities);
+        for (auto const& majority : oldMajorities)
+        {
+            if (majority.getFieldH256 (sfAmendment) == amendment)
+            {
+                if (gotMajority)
+                    return tefALREADY;
+                found = true;
+            }
+            else
+            {
+                // pass through
+                newMajorities.push_back (majority);
+            }
+        }
+    }
+
+    if (! found && lostMajority)
+        return tefALREADY;
+
+    if (gotMajority)
+    {
+        // This amendment now has a majority
+        newMajorities.push_back (STObject (sfMajority));
```

Comment:
Informational: If you use `emplace_back` you don't need the `STObject`, you can just say `newMajorities.emplace_back(sfMajority)`. Fine as is tho.